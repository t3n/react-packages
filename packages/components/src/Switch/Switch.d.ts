import React from 'react';
import { MarginProps, WidthProps } from 'styled-system';
import { ThemeFeedbackColor } from '@t3n/theme/src/theme/colors/colors';
export declare type VariantType = 'light' | 'dark';
export interface SwitchProps extends MarginProps, WidthProps {
    label?: string;
    name: string;
    value: any;
    checked: boolean;
    disabled?: boolean;
    variant?: VariantType;
    feedbackColor?: ThemeFeedbackColor;
    readOnly?: boolean;
    onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}
declare const Switch: React.FC<SwitchProps>;
export default Switch;
