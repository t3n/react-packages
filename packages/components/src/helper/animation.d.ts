import { ThemeProps } from '@t3n/theme';
export declare const backgroundAnimation: import("styled-components").FlattenInterpolation<import("styled-components").ThemedStyledProps<ThemeProps, any>>;
