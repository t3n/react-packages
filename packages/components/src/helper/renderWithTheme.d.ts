import 'jest-styled-components';
import React from 'react';
import { RenderOptions } from '@testing-library/react';
import { Theme } from '@t3n/theme';
interface OptionalTheme {
    theme?: Theme;
}
export declare const renderWithTheme: (ui: React.ReactElement, { theme, ...options }: RenderOptions & OptionalTheme) => import("@testing-library/react").RenderResult<typeof import("@testing-library/dom/types/queries"), HTMLElement>;
export {};
