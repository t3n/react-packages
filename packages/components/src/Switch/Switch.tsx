/* eslint-disable no-nested-ternary */
import React from 'react';
import styled from 'styled-components';
import { MarginProps, space, variant, width, WidthProps } from 'styled-system';

import { ThemeFeedbackColor } from '@t3n/theme/src/theme/colors/colors';

import Box from '../Box';
import Text from '../Text';

export type VariantType = 'light' | 'dark';

export interface SwitchProps extends MarginProps, WidthProps {
  label?: string;
  name: string;
  value: any;
  checked: boolean;
  disabled?: boolean;
  variant?: VariantType;
  feedbackColor?: ThemeFeedbackColor;
  readOnly?: boolean;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const StyledSwitch = styled(Box)<Omit<SwitchProps, 'name' | 'value'>>`
  display: inline-block;
  position: relative;
  min-width: 2.5rem;
  width: 2.5rem;
  height: 1.5rem;
  border-radius: 0.75rem;
  transition: all 0.1s ease-in-out;
  ${({ theme }) => space({ ml: 3, theme })}

  ${({ checked, disabled }) =>
    variant({
      variants: {
        light: {
          bg: disabled
            ? 'shades.grey232'
            : checked
              ? 'feedback.success'
              : checked && disabled
                ? 'shades.grey232'
                : 'shades.grey204',
          border: '2px solid',
          borderColor: disabled
            ? 'shades.grey232'
            : checked && disabled
              ? 'shades.grey232'
              : checked
                ? 'feedback.success'
                : 'shades.grey204',
        },
        dark: {
          bg: disabled
            ? 'shades.black'
            : checked
              ? 'feedback.success'
              : checked && disabled
                ? 'shades.black'
                : 'shades.grey95',
          border: '2px solid',
          borderColor: disabled
            ? 'shades.black'
            : checked && disabled
              ? 'shades.black'
              : checked
                ? 'feedback.success'
                : 'shades.grey95',
        },
      },
    })}
`;

const StyledSwitchToggle = styled.div<Omit<SwitchProps, 'name' | 'value'>>`
  position: absolute;
  width: 1.5rem;
  height: 1.5rem;
  top: -2px;
  left: -2px;
  border-radius: 50%;
  transition: all 0.15s ease-in-out;
  transform: translateX(${({ checked }) => (checked ? '1rem' : 0)});

  ${({ checked, disabled }) =>
    variant({
      variants: {
        light: {
          bg: 'shades.white',
          border: '2px solid',
          borderColor: disabled
            ? 'shades.grey232'
            : disabled && checked
              ? 'shades.grey232'
              : checked
                ? 'feedback.success'
                : 'shades.grey204',
        },
        dark: {
          bg: 'shades.grey42',
          border: '2px solid',
          borderColor: disabled
            ? 'shades.black'
            : disabled && checked
              ? 'shades.black'
              : checked
                ? 'feedback.success'
                : 'shades.grey95',
        },
      },
    })}
`;

const HiddenCheckbox = styled.input.attrs({ type: 'checkbox' })<SwitchProps>`
  position: absolute;
  width: 100%;
  height: 100%;
  opacity: 0;
  margin: 0;
  cursor: ${({ disabled, readOnly }) =>
    disabled ? 'not-allowed' : readOnly ? 'default' : 'pointer'};
`;

const StyledLabel = styled.label<
  Omit<SwitchProps, 'name' | 'value' | 'checked'>
>`
  display: inline-flex;
  justify-content: space-between;
  align-items: center;
  cursor: ${({ disabled, readOnly }) =>
    disabled ? 'not-allowed' : readOnly ? 'default' : 'pointer'};
  ${width};

  ${({ disabled }) =>
    variant({
      variants: {
        light: {
          color: disabled ? 'shades.grey155' : 'black',
        },
        dark: {
          color: disabled ? 'shades.grey155' : 'white',
        },
      },
    })}
`;

const PlainSwitch = ({
  checked,
  readOnly,
  onChange,
  disabled,
  name,
  variant: variantProp,
  value,
}: SwitchProps) => {
  return (
    <StyledSwitch variant={variantProp} checked={checked} disabled={disabled}>
      <StyledSwitchToggle
        variant={variantProp}
        checked={checked}
        disabled={disabled}
      />
      <HiddenCheckbox
        checked={checked}
        onChange={readOnly ? undefined : onChange}
        readOnly={readOnly}
        disabled={disabled}
        name={name}
        value={value}
      />
    </StyledSwitch>
  );
};

const Switch: React.FC<SwitchProps> = ({
  checked,
  readOnly,
  onChange,
  label,
  disabled,
  variant: variantProp,
  name,
  value,
  width: widthProp,
  ...marginProps
}) => {
  return (
    <StyledLabel
      variant={variantProp}
      disabled={disabled}
      readOnly={readOnly}
      width={widthProp}
      {...marginProps}
    >
      <Text inline>{label}</Text>
      <PlainSwitch
        variant={variantProp}
        checked={checked}
        disabled={disabled}
        onChange={readOnly ? undefined : onChange}
        readOnly={readOnly}
        name={name}
        value={value}
      />
    </StyledLabel>
  );
};

Switch.defaultProps = {
  variant: 'light',
};

export default Switch;
