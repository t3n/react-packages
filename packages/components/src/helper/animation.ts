import { css, keyframes } from 'styled-components';

import { ThemeProps } from '@t3n/theme';

const backgroundAnimationKeyframes = ({ theme }: ThemeProps) => keyframes`
  0% {
    background-color: ${theme.colors.shades.grey204}
  }

  50% {
    background-color: ${theme.colors.shades.grey244}
  }

  100% {
    background-color: ${theme.colors.shades.grey204}
  }
`;

// eslint-disable-next-line import/prefer-default-export
export const backgroundAnimation = css`
  animation: ${backgroundAnimationKeyframes} 2s linear infinite;
`;
