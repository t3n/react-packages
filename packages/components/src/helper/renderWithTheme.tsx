/* eslint-disable import/no-extraneous-dependencies */
import 'jest-styled-components';

import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { Theme, theme as originalTheme } from '@t3n/theme';

interface OptionalTheme {
  theme?: Theme;
}

// eslint-disable-next-line import/prefer-default-export
export const renderWithTheme = (
  ui: React.ReactElement,
  { theme, ...options }: RenderOptions & OptionalTheme
) => {
  // merge over default theme

  const mergedTheme = { ...originalTheme, ...theme };

  const Wrapper: React.FC = ({ children }) => (
    <ThemeProvider theme={mergedTheme}>{children}</ThemeProvider>
  );

  return render(ui, { wrapper: Wrapper, ...options });
};
