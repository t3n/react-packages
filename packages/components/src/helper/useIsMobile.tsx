import { Theme } from '@t3n/theme';
import { useCallback, useContext, useEffect, useState } from 'react';
import { ThemeContext } from 'styled-components';

const useIsMobile = (): boolean => {
  const [isMobile, setIsMobile] = useState(false);

  const theme: Theme = useContext(ThemeContext);
  const fontSize = window
    .getComputedStyle(window.document.body)
    .getPropertyValue('font-size');
  const mobileBreakpoint =
    parseInt(theme.breakpoints[1], 10) * parseInt(fontSize, 10);

  const testIsMobile = useCallback(() => {
    if (window.matchMedia(`(max-width: ${mobileBreakpoint}px)`).matches) {
      setIsMobile(true);
    } else {
      setIsMobile(false);
    }
  }, [mobileBreakpoint]);

  useEffect(() => {
    testIsMobile();
    window.addEventListener('resize', testIsMobile);

    return () => {
      window.removeEventListener('resize', testIsMobile);
    };
  }, [testIsMobile]);

  return isMobile;
};

export default useIsMobile;
