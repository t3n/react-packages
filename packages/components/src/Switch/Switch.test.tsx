import React from 'react';

import { renderWithTheme } from '../helper/renderWithTheme';
import { Switch } from './Switch';

test('readonly is set as HTML Attribute', () => {
  const { getByRole } = renderWithTheme(
    <Switch
      disabled={false}
      checked={false}
      value="true"
      name="Read-Only"
      readOnly
      label="Switch read only"
      variant="light"
    />,
    {}
  );

  expect(getByRole('checkbox').hasAttribute('readonly')).toBeTruthy();
});

test('Switch variants matches snapshot', () => {
  const { container } = renderWithTheme(
    <>
      <Switch
        disabled={false}
        checked={false}
        value="true"
        name="unchecked"
        readOnly
        label="Switch unchecked"
        variant="light"
      />
      <Switch
        disabled={false}
        checked
        value="true"
        name="checked"
        readOnly
        label="Switch checked"
        variant="light"
      />
      <Switch
        disabled
        checked
        value="true"
        name="disabled"
        readOnly
        label="Switch disabled checked"
        variant="light"
      />
      <Switch
        disabled
        checked={false}
        value="true"
        name="disabled"
        readOnly
        label="Switch disabled unchecked"
        variant="light"
      />
      <Switch
        disabled={false}
        checked={false}
        value="true"
        name="unchecked"
        readOnly
        label="Switch unchecked"
        variant="dark"
      />
      <Switch
        disabled={false}
        checked
        value="true"
        name="checked"
        readOnly
        label="Switch checked"
        variant="dark"
      />
      <Switch
        disabled
        checked
        value="true"
        name="disabled"
        readOnly
        label="Switch disabled checked"
        variant="dark"
      />
      <Switch
        disabled
        checked={false}
        value="true"
        name="disabled"
        readOnly
        label="Switch disabled unchecked"
        variant="dark"
      />
    </>,
    {}
  );
  expect(container).toMatchSnapshot();
});
