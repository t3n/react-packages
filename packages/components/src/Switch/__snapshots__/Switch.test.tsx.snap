// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch variants matches snapshot 1`] = `
.c1 {
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 1rem;
  line-height: 1.5em;
}

.c2 {
  display: inline-block;
  position: relative;
  min-width: 2.5rem;
  width: 2.5rem;
  height: 1.5rem;
  border-radius: 0.75rem;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  margin-left: 16px;
  background-color: #cccccc;
  border: 2px solid;
  border-color: #cccccc;
}

.c5 {
  display: inline-block;
  position: relative;
  min-width: 2.5rem;
  width: 2.5rem;
  height: 1.5rem;
  border-radius: 0.75rem;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  margin-left: 16px;
  background-color: #34e88f;
  border: 2px solid;
  border-color: #34e88f;
}

.c8 {
  display: inline-block;
  position: relative;
  min-width: 2.5rem;
  width: 2.5rem;
  height: 1.5rem;
  border-radius: 0.75rem;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  margin-left: 16px;
  background-color: #e8e8e8;
  border: 2px solid;
  border-color: #e8e8e8;
}

.c13 {
  display: inline-block;
  position: relative;
  min-width: 2.5rem;
  width: 2.5rem;
  height: 1.5rem;
  border-radius: 0.75rem;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  margin-left: 16px;
  background-color: #8F8F8F;
  border: 2px solid;
  border-color: #8F8F8F;
}

.c16 {
  display: inline-block;
  position: relative;
  min-width: 2.5rem;
  width: 2.5rem;
  height: 1.5rem;
  border-radius: 0.75rem;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  margin-left: 16px;
  background-color: #000;
  border: 2px solid;
  border-color: #000;
}

.c3 {
  position: absolute;
  width: 1.5rem;
  height: 1.5rem;
  top: -2px;
  left: -2px;
  border-radius: 50%;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  background-color: #fff;
  border: 2px solid;
  border-color: #cccccc;
}

.c6 {
  position: absolute;
  width: 1.5rem;
  height: 1.5rem;
  top: -2px;
  left: -2px;
  border-radius: 50%;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  -webkit-transform: translateX(1rem);
  -ms-transform: translateX(1rem);
  transform: translateX(1rem);
  background-color: #fff;
  border: 2px solid;
  border-color: #34e88f;
}

.c9 {
  position: absolute;
  width: 1.5rem;
  height: 1.5rem;
  top: -2px;
  left: -2px;
  border-radius: 50%;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  -webkit-transform: translateX(1rem);
  -ms-transform: translateX(1rem);
  transform: translateX(1rem);
  background-color: #fff;
  border: 2px solid;
  border-color: #e8e8e8;
}

.c11 {
  position: absolute;
  width: 1.5rem;
  height: 1.5rem;
  top: -2px;
  left: -2px;
  border-radius: 50%;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  background-color: #fff;
  border: 2px solid;
  border-color: #e8e8e8;
}

.c14 {
  position: absolute;
  width: 1.5rem;
  height: 1.5rem;
  top: -2px;
  left: -2px;
  border-radius: 50%;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  background-color: #2a2a2a;
  border: 2px solid;
  border-color: #8F8F8F;
}

.c15 {
  position: absolute;
  width: 1.5rem;
  height: 1.5rem;
  top: -2px;
  left: -2px;
  border-radius: 50%;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  -webkit-transform: translateX(1rem);
  -ms-transform: translateX(1rem);
  transform: translateX(1rem);
  background-color: #2a2a2a;
  border: 2px solid;
  border-color: #34e88f;
}

.c17 {
  position: absolute;
  width: 1.5rem;
  height: 1.5rem;
  top: -2px;
  left: -2px;
  border-radius: 50%;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  -webkit-transform: translateX(1rem);
  -ms-transform: translateX(1rem);
  transform: translateX(1rem);
  background-color: #2a2a2a;
  border: 2px solid;
  border-color: #000;
}

.c18 {
  position: absolute;
  width: 1.5rem;
  height: 1.5rem;
  top: -2px;
  left: -2px;
  border-radius: 50%;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  background-color: #2a2a2a;
  border: 2px solid;
  border-color: #000;
}

.c4 {
  position: absolute;
  width: 100%;
  height: 100%;
  opacity: 0;
  margin: 0;
  cursor: default;
}

.c10 {
  position: absolute;
  width: 100%;
  height: 100%;
  opacity: 0;
  margin: 0;
  cursor: not-allowed;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: default;
  color: black;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: not-allowed;
  color: shades.grey155;
}

.c12 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: default;
  color: white;
}

@media screen and (min-width:30rem) {
  .c1 {
    line-height: 1.5em;
  }
}

@media screen and (min-width:48rem) {
  .c1 {
    line-height: 1.666em;
  }
}

<div>
  <label
    class="c0"
    readonly=""
  >
    <span
      class="c1"
    >
      Switch unchecked
    </span>
    <div
      class="c2"
    >
      <div
        class="c3"
      />
      <input
        class="c4"
        name="unchecked"
        readonly=""
        type="checkbox"
        value="true"
      />
    </div>
  </label>
  <label
    class="c0"
    readonly=""
  >
    <span
      class="c1"
    >
      Switch checked
    </span>
    <div
      class="c5"
    >
      <div
        class="c6"
      />
      <input
        checked=""
        class="c4"
        name="checked"
        readonly=""
        type="checkbox"
        value="true"
      />
    </div>
  </label>
  <label
    class="c7"
    disabled=""
    readonly=""
  >
    <span
      class="c1"
    >
      Switch disabled checked
    </span>
    <div
      class="c8"
      disabled=""
    >
      <div
        class="c9"
        disabled=""
      />
      <input
        checked=""
        class="c10"
        disabled=""
        name="disabled"
        readonly=""
        type="checkbox"
        value="true"
      />
    </div>
  </label>
  <label
    class="c7"
    disabled=""
    readonly=""
  >
    <span
      class="c1"
    >
      Switch disabled unchecked
    </span>
    <div
      class="c8"
      disabled=""
    >
      <div
        class="c11"
        disabled=""
      />
      <input
        class="c10"
        disabled=""
        name="disabled"
        readonly=""
        type="checkbox"
        value="true"
      />
    </div>
  </label>
  <label
    class="c12"
    readonly=""
  >
    <span
      class="c1"
    >
      Switch unchecked
    </span>
    <div
      class="c13"
    >
      <div
        class="c14"
      />
      <input
        class="c4"
        name="unchecked"
        readonly=""
        type="checkbox"
        value="true"
      />
    </div>
  </label>
  <label
    class="c12"
    readonly=""
  >
    <span
      class="c1"
    >
      Switch checked
    </span>
    <div
      class="c5"
    >
      <div
        class="c15"
      />
      <input
        checked=""
        class="c4"
        name="checked"
        readonly=""
        type="checkbox"
        value="true"
      />
    </div>
  </label>
  <label
    class="c7"
    disabled=""
    readonly=""
  >
    <span
      class="c1"
    >
      Switch disabled checked
    </span>
    <div
      class="c16"
      disabled=""
    >
      <div
        class="c17"
        disabled=""
      />
      <input
        checked=""
        class="c10"
        disabled=""
        name="disabled"
        readonly=""
        type="checkbox"
        value="true"
      />
    </div>
  </label>
  <label
    class="c7"
    disabled=""
    readonly=""
  >
    <span
      class="c1"
    >
      Switch disabled unchecked
    </span>
    <div
      class="c16"
      disabled=""
    >
      <div
        class="c18"
        disabled=""
      />
      <input
        class="c10"
        disabled=""
        name="disabled"
        readonly=""
        type="checkbox"
        value="true"
      />
    </div>
  </label>
</div>
`;
